### ISSUE TITLE
Use original file paths

### ISSUE THREAD

**Issue by theferrit32** [common-workflow-language/cwltool#559]
Is there any way within the existing cwltool to tell it to just perform a cwl job with the original file paths it contains rather than creating and linking to temp directories?  My use case involves executing cwltool in a remote environment that contains consistently mounted remote filesystems, and it is complicated by cwltool's changing of file paths to ones generated at runtime.

**Comment by mr-c** [comment]
Hello @theferrit32 

Is your need related to debugging, or is there an issue with making copies?

If the later there is a cwltool specific extension that allows Files and Directories to be edited in place:
https://github.com/common-workflow-language/user_guide/issues/49

What you describe reminds me of my experience converting tools and workflows from a system where a unified filesystem is available on all nodes and the various scripts and logics rely on that fact. Is that the case for you?

**Comment by LourensVeen** [comment]
If you control the complete CWL definition, then you could pass in your files as strings, rather than as File objects. If your output file names are not exactly known, then I think you could still glob them and return the path attribute of the File object that gets you, using an Expression.

**Comment by mr-c** [comment]
FYI: What @LourensVeen has proposed leads to non-portable workflows that won't work on systems without a shared filesystem. I can not recommend that approach at all.

**Comment by LourensVeen** [comment]
True, but without the shared filesystem you'll have different results anyway, right? Because you don't have in-place update. Maybe I don't understand the problem. @theferrit32, can you give a bit more background?

**Comment by theferrit32** [comment]
The use case is to have cwltool parse the cwl document and the input values for the job, and open a process that will handle the execution of the workflow.  However the execution is done in a parallelized and asynchronous way on a remote host, so the files specified in the job file (like myjob.yml) will most likely not be accessible by the cwltool program.  

Really my goal was to have cwltool parse all the input files and use the command line bindings to generate command strings, without making any filesystem access attempts during that process.  The topic of filesystem or path portability is not a concern to us, because we are already dealing with a shared filesystem mounted in a defined manner.  That mountpoint being set up correctly on all nodes can be assumed to be a requirement of being able to run the workflow. 

I have worked around this for now by subclassing CommandLineTool and overriding behavior that accesses the filesystem, but left in the existing behavior so it is not broken if the user wants to use the regular cwltool.  The only reason why I did not want to go with the purely string specification of file paths, and wanted to stay with fields of class/type: File was so that a job file created for the standard cwltool can work both with the original cwltool method as intended, and with my method.

**Comment by tetron** [comment]
That sounds like a reasonable solution.  The current recommended way to adapt cwltool to various environments is to created a derived version which overrides the path handling and job dispatch behavior (for example, arvados-cwl-runner, cwl-tes).  

What scheduler are you using?  If there is a public repo we can link to it on the front page.

**Comment by theferrit32** [comment]
I am using using Chronos in a mesos cluster.  The goal is maximal parallelization across nodes, by looking at each workflow step and grouping them together in sets of jobs that can run at the same time, and then looking at input/output dependencies and enforcing a strict order of the jobs. 

The fork this is currently being developed in is here: https://github.com/stevencox/cwltool

Feel free to take a look.  It's currently very dependent on some hardcoded configurations I am using during early development so it is not all that generally useful, but portability is being worked on.
